DirectXTex/DirectXTexMipmaps.cpp

```diff
@@ -2521,6 +2521,129 @@ namespace
+
+    for (size_t h = 0; h < image.height - 1; ++h)
+    {
+        if (!_LoadScanline(scanline.get(), image.width, pPixels, image.rowPitch, image.format))
+            continue;
+
+        XMVECTOR* currentRowPixels = scanline.get();
+
+        if (!_LoadScanline(scanline.get(), image.width, pPixels  + image.rowPitch, image.rowPitch, image.format))
+            continue;
+
+        XMVECTOR* nextRowPixels = scanline.get();
+
+        for (size_t w = 0; w < image.width - 1; ++w)
+        {            
+            // TODO: Use std::clamp when c++17 is available.
+            float alpha00 = clamp(currentRowPixels[w].m128_f32[3] * scale, 0.0f, 1.0f);
+            float alpha10 = clamp(currentRowPixels[w + 1].m128_f32[3] * scale, 0.0f, 1.0f);
+            float alpha01 = clamp(nextRowPixels[w].m128_f32[3] * scale, 0.0f, 1.0f);
+            float alpha11 = clamp(nextRowPixels[w + 1].m128_f32[3] * scale, 0.0f, 1.0f);
+
+            for (float fy = 0.5f / n; fy < 1.0f; fy++) {
+                for (float fx = 0.5f / n; fx < 1.0f; fx++) {
+                    float alpha = alpha00 * (1 - fx) * (1 - fy) + alpha10 * fx * (1 - fy) + alpha01 * (1 - fx) * fy + alpha11 * fx * fy;
+                    if (alpha > alphaCoverageRatio)
+                    {
+                        coverage += 1.0f;
+                    }
+                }
+            }
+        }
+
+        pPixels += image.rowPitch;
+    }
+
+    return coverage / float(width * height * n * n);
+}
+
+_Use_decl_annotations_
+void ScaleAlphaToCoverage(Image& image, const float alphaCoverageRatio, const float desiredCoverage)
+{
+    float minScale = 0.0f;
+    float maxScale = 4.0f;
+    float scale = 1.0f;
+
+    static const int s_iterationCount = 10;
+    for (int scalingIndex = 0; scalingIndex < s_iterationCount; ++scalingIndex)
+    {
+        const float curCoverage = CalculateAlphaCoverage(image, alphaCoverageRatio, scale);
+        if (curCoverage < desiredCoverage)
+        {
+            minScale = scale;
+        }
+        else if (curCoverage > desiredCoverage)
+        {
+            maxScale = scale;
+        }
+        scale = (minScale + maxScale) * 0.5f;
+    }
+
+    // No scaling required
+    if (scale == 1.0f)
+    {
+        return;
+    }
+
+    // Find closest scale
+    const float minVarience = abs(desiredCoverage - CalculateAlphaCoverage(image, alphaCoverageRatio, minScale));
+    const float curVarience = abs(desiredCoverage - CalculateAlphaCoverage(image, alphaCoverageRatio, scale));
+    const float maxVarience = abs(desiredCoverage - CalculateAlphaCoverage(image, alphaCoverageRatio, maxScale));
+
+    if (minVarience < curVarience && minVarience < maxVarience)
+    {
+        scale = minScale;
+    }
+    else if (maxVarience < minVarience && maxVarience < curVarience)
+    {
+        scale = maxScale;
+    }
+
+    ScopedAlignedArrayXMVECTOR scanline(reinterpret_cast<XMVECTOR*>(_aligned_malloc((sizeof(XMVECTOR) * image.width), 16)));
```

Comment:
As I noted above, I'd pass the allocated memory buffer to ``CalculateAlphaCoverage`` and then reuse it here.